packages/opentelemetry-plugin-express/src/express.ts

```diff
@@ -0,0 +1,220 @@
+   * @param original
+   */
+  private _getAppUsePatch(
+    original: core.ApplicationRequestHandler<express.Application>
+  ) {
+    const plugin = this;
+    return function use(
+      this: { _router: ExpressRouter },
+      ...args: Parameters<typeof original>
+    ) {
+      const route = original.apply(this, args);
+      const layer = this._router.stack[this._router.stack.length - 1];
+      plugin._applyPatch(
+        layer,
+        typeof args[0] === 'string' ? args[0] : undefined
+      );
+      return route;
+      // tslint:disable-next-line:no-any
+    } as any;
+  }
+
+  /** Patch each express layer to create span and propagate scope */
+  private _applyPatch(layer: ExpressLayer, layerPath?: string) {
+    const plugin = this;
+    if (layer[kLayerPatched] === true) return;
+    layer[kLayerPatched] = true;
+    this._logger.debug('patching express.Router.Layer.handle');
+    shimmer.wrap(layer, 'handle', function(original: Function) {
+      if (original.length === 4) return original;
+
+      return function(
+        this: ExpressLayer,
+        req: PatchedRequest,
+        res: express.Response,
+        next: express.NextFunction
+      ) {
+        storeLayerPath(req, layerPath);
+        const route = (req[_LAYERS_STORE_PROPERTY] as string[]).join('');
+        const attributes: Attributes = {
+          [AttributeNames.COMPONENT]: plugin._COMPONENT,
+          [AttributeNames.HTTP_ROUTE]: route.length > 0 ? route : undefined,
+        };
+        const metadata = getLayerMetadata(layer, layerPath);
+        const type = metadata.attributes[
+          AttributeNames.EXPRESS_TYPE
+        ] as ExpressLayerType;
+        // verify against the config if the layer should be ignored
+        if (isLayerIgnored(metadata.name, type, plugin._config)) {
+          return original.apply(this, arguments);
+        }
+        const span = plugin._tracer.startSpan(metadata.name, {
+          parent: plugin._tracer.getCurrentSpan(),
+          attributes: Object.assign(attributes, metadata.attributes),
+        });
+        // verify we have a callback
+        let callbackIdx = Array.from(arguments).findIndex(
+          arg => typeof arg === 'function'
+        );
+        let callbackHasBeenCalled = false;
+        if (callbackIdx >= 0) {
+          arguments[callbackIdx] = function() {
+            callbackHasBeenCalled = true;
+            if (!(req.route && arguments[0] instanceof Error)) {
+              (req[_LAYERS_STORE_PROPERTY] as string[]).pop();
+            }
+            return patchEnd(span, plugin._tracer.bind(next))();
+          };
+        }
+        const result = original.apply(this, arguments);
+        // if the layer return a response, the callback will never
+        // be called, so we need to manually close the span
+        if (callbackHasBeenCalled === false) {
+          span.end();
+        }
+        return result;
+      };
+    });
+  }
+}
+
+export const plugin = new ExpressPlugin('express');
```

Comment:
nit: `new ExpressPlugin(ExpressPlugin.component)`. Although this might be overkill :)